ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -613,24 +606,33 @@ private synchronized CompletableFuture<Void> changeToFollowerAsync(
     return future;
   }
 
-  synchronized void changeToFollowerAndPersistMetadata(
+  synchronized CompletableFuture<Void> changeToFollowerAndPersistMetadata(
       long newTerm,
       boolean allowListener,
       Object reason) throws IOException {
-    if (changeToFollower(newTerm, false, allowListener, reason)) {
-      state.persistMetadata();
+    final AtomicBoolean metadataUpdated = new AtomicBoolean();
+    final CompletableFuture<Void> future = changeToFollower(newTerm, false, allowListener, reason, metadataUpdated);
+    try {
+      if (metadataUpdated.get()) {
+        state.persistMetadata();
+      }
+    } catch (IOException e) {
+      CompletableFuture.runAsync(future::join);
```

Comment:
`CompletableFuture.runAsync` means use another thread to call `join()`.  It is not useful.  More details:
- Without this line, the shutdown thread will run and complete the future.
- With this line, the shutdown thread will run and complete the future.  In addition, we use another thread to wait for the shutdown thread.  Why use one more thread?

Reply:
Assuming we encounter an exception here, we either do not handle the future or need another thread to handle it. If we do not use another thread to handle the future, there is still a deadlock problem of holding the lock and waiting for the future.

The same consideration applies to using runAsync elsewhere

Reply:
If we remove line 620, how could we have a deadlock?  Could you give more details?

Reply:
Replacing CompleteFuture.runAsync(future::join) with future.join still presents the deadlock issue where thread A holds the lock while waiting for the future, and thread B needs to acquire the lock.

If we simply ignore the future and throw an exception without handling it, then no deadlock will occur.

Reply:
> Replacing CompleteFuture.runAsync(future::join) with future.join ...

I mean "remove the line", not "replace it with future.join", i.e. case 3 below.

Three cases:
1. With `future.join()`, the current thread will wait for the shutdown thread to complete
2. With `CompleteFuture.runAsync(future::join)`, the current thread and the shutdown thread will run in parallel.  In addition, it uses one more thread to wait for the shutdown thread. 
3. With an empty line, the current thread and the shutdown thread will run in parallel.

Reply:
I see. 

case 1. We should avoid this situation, which seems to contradict our issue
case 2. This writing style does not seem to be recommended
case 3. It seems that we can only use this

Then we can use the third case